### PR TITLE
Handle scientific notation when parsing sizes

### PR DESCRIPTION
**Pull request by raxbg**
Parsing numbers written in scientific notation raises `UnexpectedTokenException`. This PR fixes this.

It might be a good idea to separate the number parsing and rendering from the `Size` class, however this will break backwards compatibility, so it might have to wait for a new major version release. Otherwise the numbers will have to extend `Size`...

### CODE REVIEW

**Comment by sabberworm** [comment]
> It might be a good idea to separate the number parsing and rendering from the `Size` class, however this will break backwards compatibility, so it might have to wait for a new major version release. Otherwise the numbers will have to extend `Size`...

I’m not exactly sure what you mean. Size is exactly designed to handle numbers… But I agree that we should probably store additional metadata about how the size was formatted originally and provide ways for `OutputFormat` to determine if the output should be normalized (and to what) or output as it was input.

File: lib/Sabberworm/CSS/Value/Size.php [hunk 1]
```diff
@@ -28,9 +28,11 @@ public static function parse(ParserState $oParserState, $bIsColorComponent = fal
 		if ($oParserState->comes('-')) {
 			$sSize .= $oParserState->consume('-');
 		}
-		while (is_numeric($oParserState->peek()) || $oParserState->comes('.')) {
+		while (is_numeric($oParserState->peek()) || $oParserState->comes('.') || $oParserState->comes('e+', true) || $oParserState->comes('e-', true)) {
```
Comment (sabberworm) [inline lib/Sabberworm/CSS/Value/Size.php]: If I’m not mistaken, the `+` isn’t mandatory.

**Comment by raxbg** [comment]
I was thinking of a `Number` class that can be used to parse/output numbers. `Size` can use `Numbers` to store the numeric portion of its numbers. Right now `Size` is a little weird, because it holds all kinds of numbers even colors instead of having a `Color` that holds `Number` objects for its numeric parts. Having this further separation will make it simpler for third parties when working with plain numbers, sizes and colors generated by the parser.

File: lib/Sabberworm/CSS/Value/Size.php (hunk 1 above)
Comment (raxbg) [reply lib/Sabberworm/CSS/Value/Size.php]: True. Fixed.

File: tests/files/scientific-notation-numbers.css [hunk 2]
```diff
@@ -0,0 +1,6 @@
+body {
+  background-color: rgba(62,174,151,3.0418206565232E+21);
+  z-index: 3.0418206565232E-2;
+  font-size: 1em;
+  top: 1.923478e2px;
```
Comment (sabberworm) [inline tests/files/scientific-notation-numbers.css]: Codacy Issue found: [Unexpected token '1.923478e2px' at line 5, col 8.](https://app.codacy.com/manual/sabberworm/PHP-CSS-Parser/pullRequest?prid=5203547)

**Comment by sabberworm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- lib/Sabberworm/CSS/Value/Size.php  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sabberworm/PHP-CSS-Parser/pullRequest?prid=5203547&bid=17067897)

File: tests/files/scientific-notation-numbers.css [hunk 3]
```diff
@@ -0,0 +1,6 @@
+body {
+  background-color: rgba(62,174,151,3.0418206565232E+21);
```
Comment (sabberworm) [inline tests/files/scientific-notation-numbers.css]: Codacy Issue found: [Expected RBRACE at line 2, col 37.](https://app.codacy.com/manual/sabberworm/PHP-CSS-Parser/pullRequest?prid=5203547)

File: tests/files/scientific-notation-numbers.css [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+  z-index: 3.0418206565232E-2;
```
Comment (sabberworm) [inline tests/files/scientific-notation-numbers.css]: Codacy Issue found: [Unexpected token '3.0418206565232E-2' at line 3, col 12.](https://app.codacy.com/manual/sabberworm/PHP-CSS-Parser/pullRequest?prid=5203547)

**Pull request merged**